java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java

```diff
@@ -0,0 +1,189 @@
+    }
+    return bucket;
+  }
+
+  @Override
+  public void visit(Visitor visitor)
+      throws IOException {
+    traverse(visitor, new DictionaryUtils.VisitorContextImpl(this.byteArray, this.keyOffsets));
+  }
+
+  private void traverse(Visitor visitor, DictionaryUtils.VisitorContextImpl context) throws IOException {
+    for (DynamicIntArray intArray : hashArray) {
+      for (int i = 0; i < intArray.size() ; i ++) {
+        context.setPosition(intArray.get(i));
+        visitor.visit(context);
+      }
+    }
+  }
+
+  @Override
+  public void clear() {
+    byteArray.clear();
+    keyOffsets.clear();
+    Arrays.fill(hashArray, null);
+  }
+
+  @Override
+  public void getText(Text result, int position) {
+    DictionaryUtils.getTextInternal(result, position, this.keyOffsets, this.byteArray);
+  }
+
+  @Override
+  public int add(byte[] bytes, int offset, int length) {
+    resizeIfNeeded();
+    newKey.set(bytes, offset, length);
+    return add(newKey);
+  }
+
+  public int add(Text text) {
+    resizeIfNeeded();
+
+    int index = getIndex(text);
+    DynamicIntArray candidateArray = hashArray[index];
+
+    newKey.set(text);
+
+    Text tmpText = new Text();
+    for (int i = 0; i < candidateArray.size(); i++) {
+      getText(tmpText, candidateArray.get(i));
+      if (tmpText.equals(newKey)) {
+        return candidateArray.get(i);
+      }
+    }
+
+    // if making it here, it means no match.
+    int len = newKey.getLength();
+    int currIdx = keyOffsets.size();
+    keyOffsets.add(byteArray.add(newKey.getBytes(), 0, len));
+    candidateArray.add(currIdx);
+    return currIdx;
+  }
+
+  private void resizeIfNeeded() {
+    if (keyOffsets.size() >= threshold) {
+      int oldCapacity = keyOffsets.size();
+      int newCapacity = (oldCapacity << 1) + 1;
+      doResize(newCapacity);
+      this.threshold = (int)Math.min(newCapacity * loadFactor, MAX_ARRAY_SIZE + 1);
+    }
+  }
+
+  @Override
+  public int size() {
+    return keyOffsets.size();
+  }
+
+  /**
+   * Compute the hash value and find the corresponding index.
+   *
+   */
+  int getIndex(Text text) {
+    return (text.hashCode() & 0x7FFFFFFF) % capacity;
+  }
+
+  // Resize the hash table, re-hash all the existing keys.
+  // byteArray and keyOffsetsArray don't have to be re-filled.
+  private void doResize(int newSize) {
+    DynamicIntArray[] resizedHashArray = new DynamicIntArray[newSize];
+    for (int i = 0; i < newSize; i++) {
+      resizedHashArray[i] = new DynamicIntArray();
+    }
+
+    Text tmpText = new Text();
+    for (int i = 0; i < capacity; i++) {
+      DynamicIntArray intArray = hashArray[i];
+      int bucketSize = intArray.size();
+      if (bucketSize > 0) {
+        for (int j = 0; j < bucketSize; j++) {
+          getText(tmpText, intArray.get(j));
+          int newIndex = getIndex(tmpText);
+          resizedHashArray[newIndex].add(intArray.get(j));
+        }
+      }
+    }
+
+    Arrays.fill(hashArray, null);
```

Comment:
I was thinking to ensure everything to be null just in case that some elements in the array is still be referenced so that the space doesn't get garbage-collected. Do you think that's a fair ? I don't have strong opinion about it though.